feat(blocks): add Replicate model blocks

snehaoladri opened the pull request:
This PR adds two new blocks for running Replicate models in AutoGPT:

ReplicateModelBlock: Synchronous execution of any Replicate model with custom inputs

Key Features:

Support for any public Replicate model via model name (e.g., "stability-ai/stable-diffusion")
Custom input parameters via dictionary (e.g., {"prompt": "a beautiful landscape"})
Optional model version specification for reproducible results
Proper credentials handling using ProviderName.REPLICATE
Comprehensive test suite with 12 test cases covering success, error, and edge cases
Type-safe implementation with full pyright compliance
Mock methods for testing and development

# Checklist 📋
## For code changes:
- [X] I have clearly listed my changes in the PR description
- [X] I have made a test plan
- [X] I have tested my changes according to the test plan:
  - Unit tests for ReplicateModelBlock (6 test cases)
  -  Test block initialization and configuration
  -  Test mock methods for development
  -  Test error handling and edge cases
  -  Verify type safety with pyright (0 errors)
  -  Verify code formatting with Black and isort
  -  Verify linting with Ruff (0 errors)
  -  Test credentials handling with ProviderName.REPLICATE
  -  Test model version specification functionality
  -  Test both synchronous and asynchronous execution paths

AutoGPT-Agent commented:
Thank you for your detailed PR adding Replicate model blocks to AutoGPT! The code looks well-structured with comprehensive test coverage.

### What looks good:
- Excellent test coverage with 12 test cases
- Clear documentation in the code
- Good error handling and edge case coverage
- Type-safe implementation
- Proper credential handling using ProviderName.REPLICATE

### Changes needed:

1. **PR Title Format**: Your PR title needs to follow the conventional commit format. Please update it to something like:
   ```
   feat(blocks): add Replicate model blocks
   ```
   This should specify the type (feat) and scope (blocks) according to our commit convention.

2. **Minor Suggestions**:
   - Consider adding some basic usage examples in the docstrings to make it easier for developers to use these blocks
   - The mock classes at the end of the file could be moved to the test file for cleaner separation

Once you update the PR title to follow our conventional commit format, this should be ready to merge! The implementation is solid and well-tested.

snehaoladri commented:
@Torantulino has created a new PR addressing the issues. The Vercel deployment should be completed once the PR is reviewed and approved. Could a team member please review this so we can merge it promptly?

ntindle commented on autogpt_platform/backend/backend/blocks/replicate_block.py:
```diff
@@ -0,0 +1,425 @@
+    async def run(
+        self, input_data: Input, *, credentials: APIKeyCredentials, **kwargs
+    ) -> BlockOutput:
+        if isinstance(input_data, dict):
+            input_data = self.input_schema(**input_data)
```
?

snehaoladri replied:
Updated the logic

ntindle commented on autogpt_platform/backend/backend/blocks/replicate_block.py:
```diff
@@ -0,0 +1,425 @@
+    async def run(
+        self, input_data: Input, *, credentials: APIKeyCredentials, **kwargs
+    ) -> BlockOutput:
+        if isinstance(input_data, dict):
+            input_data = self.input_schema(**input_data)
```
??

snehaoladri replied:
Updated the logic to use execute function

ntindle commented on autogpt_platform/backend/backend/blocks/replicate_block.py:
```diff
@@ -0,0 +1,425 @@
+        self,
+        model_ref: str,
+        model_inputs: Dict[str, Any],
+        api_key: str,
```
```suggestion
        api_key: SecretStr,
```

snehaoladri replied:
done

ntindle commented on autogpt_platform/backend/backend/blocks/replicate_block_test.py:
```diff
@@ -0,0 +1,343 @@
+"""
```
we shouldn't need this as the block itself can have tests on it

snehaoladri replied:
removed

snehaoladri commented on autogpt_platform/backend/pyproject.toml:
```diff
@@ -71,6 +71,7 @@ aiofiles = "^24.1.0"
 tiktoken = "^0.9.0"
 aioclamd = "^1.0.0"
 setuptools = "^80.9.0"
+typing-extensions = "^4.14.0"
```
removed it

snehaoladri commented on autogpt_platform/backend/backend/blocks/replicate_block.py:
```diff
@@ -0,0 +1,425 @@
+
+    def __init__(self):
+        super().__init__(
+            id="a1b2c3d4-e5f6-7890-abcd-ef1234567890",
```
done

ntindle commented:
I'll pop this on my todo list :) waiting for some PRs to merge to fix ci then I'll take a look

AutoGPT-Agent commented:
Thank you for this PR adding Replicate model block functionality! The implementation looks solid with good error handling and type safety.

A few observations:

1. Your PR description mentions two blocks (ReplicateModelBlock and ReplicateAsyncModelBlock), but I only see the implementation for ReplicateModelBlock in the diff. Is the ReplicateAsyncModelBlock implementation missing from the PR, or was this a change in scope?

2. The code structure looks good with proper separation of concerns - moving authentication and helper functions to separate files.

3. I appreciate the comprehensive test plan and the attention to error handling in the implementation.

4. The cost configuration for the new block has been properly added to block_cost_config.py.

Once you clarify the status of the ReplicateAsyncModelBlock (whether it's coming in a separate PR or was intended to be part of this one), this PR should be good to go.
